support "header:" metadata convention

mattigruener opened the pull request:
For simplicity, this only supports strings. John, do you think we need to add the rest of the data types? For what I'm planning on doing with it, I can easily cast everything to a string earlier when I built the metadata in a Python expression. 

Thanks :)

P.S.: this will currently declare 'custom_attributes' on drivers that don't support it as well. It's not causing any harm, but it might be cleaner to only do it on exr and deepexr drivers?

johnhaddon commented on src/GafferArnold/IECoreArnoldPreview/Renderer.cpp:
```diff
@@ -176,15 +176,37 @@ class ArnoldOutput : public IECore::RefCounted
 
+			if( headerVector.size() > 0 )
+			{
+				IECore::StringVectorDataPtr headerData = new IECore::StringVectorData( headerVector );
```
Rather than copy construct headerVector here, I think we may as well just construct the StringVectorData up top, and then fill it directly :

```
StringVectorDataPtr customAttributesData = new StringVectorData();
vector<string> &customAttributes = customAttributesData->readable();
```

johnhaddon commented on src/GafferArnold/IECoreArnoldPreview/Renderer.cpp:
```diff
@@ -176,15 +176,37 @@ class ArnoldOutput : public IECore::RefCounted
 				ParameterAlgo::setParameter( m_driver.get(), it->first.c_str(), it->second.get() );
 			}
 
+			if( headerVector.size() > 0 )
```
We should remove this check and always set the parameter, so that in interactive renders we reset the Arnold parameter appropriately if the header is removed in Gaffer.

johnhaddon commented on src/GafferArnold/IECoreArnoldPreview/Renderer.cpp:
```diff
@@ -176,15 +176,37 @@ class ArnoldOutput : public IECore::RefCounted
+					const IECore::StringData *headerEntry = IECore::runTimeCast<const IECore::StringData>( it->second.get() );
+					if( headerEntry )
+					{
+						headerVector.push_back( "string " + it->first.string().substr(7, it->first.value().size()-7) + " " + headerEntry->readable().c_str() );
```
We can omit the second argument to `substr()`, since the default is to give us to the end of the string, which is what we want.

mattigruener commented:
I addressed most of your feedback in the new code. Also had to change the test as it gave me inconsistent results.

As for the different datatypes - the Arnold docs say this about `custom_attributes`:

> Allows custom attributes to be added as exr metadata. Supported types are int, float, point2, matrix and string.

I will go ahead and use the respective types where possible and map bool to int. For the vector types, would you prefer v2i and v2f to be mapped to point2 or rather have their values be transformed to strings as I will probably have to do for the other vector types anyway?

mattigruener commented:
I added support for the data types that users can add to the outputs. I am sure there's a nicer way to implement this, John? We could represent all data types as strings on the metadata side? Can we call `readable()` on `DataPtr` and make all of this a one-liner?

johnhaddon commented on src/GafferArnold/IECoreArnoldPreview/Renderer.cpp:
```diff
@@ -176,15 +176,74 @@ class ArnoldOutput : public IECore::RefCounted
 			}
 
+			// adding metadata like we do here is only supported for exr and deepexr drivers
+			if( output->getType() == "exr" || output->getType() == "deepexr" )
```
Is there a reason we can't just query if the "custom_attributes" parameter exists on the driver, rather than hardcode this assumption?

johnhaddon commented on src/GafferArnold/IECoreArnoldPreview/Renderer.cpp:
```diff
@@ -176,15 +176,74 @@ class ArnoldOutput : public IECore::RefCounted
+					}
+					else
+					{
+						IECore::msg( IECore::Msg::Warning, "IECoreArnold::Renderer", boost::format( "Can not convert '%s'. Must be of type string." ) % it->first.value() );
```
"Must be of type string" is no longer true. Lets say `Cannot convert data "name" of type "type"`.

johnhaddon commented on src/GafferArnold/IECoreArnoldPreview/Renderer.cpp:
```diff
@@ -176,15 +176,74 @@ class ArnoldOutput : public IECore::RefCounted
+				std::string formattedData;
+				if( boost::starts_with( it->first.string(), "header:" ) )
+				{
+					if( const IECore::BoolData *data = IECore::runTimeCast<const IECore::BoolData>( it->second.get() ) )
```
This function is becoming huuuge. How about factoring out the big list of ifs into a `formatHeaderParameter( const Data *data )` function?

mattigruener commented:
I believe I have addressed all of your your feedback, John. Thanks :) Also added a test to make sure that the parameter isn't declared on drivers that don't support it. The warning gets tested as well now.

johnhaddon commented:
Thanks Matti - could you squash those in and then we can merge?

mattigruener commented:
Done. Thanks for your help with this.

johnhaddon commented:
Nice one - thanks Matti!

mattigruener commented on src/GafferArnold/IECoreArnoldPreview/Renderer.cpp:
```diff
@@ -176,15 +225,33 @@ class ArnoldOutput : public IECore::RefCounted
 
+			if( AiNodeEntryLookUpParameter( AiNodeGetNodeEntry( m_driver.get() ), "custom_attributes" ) )
+			{
+				ParameterAlgo::setParameter( m_driver.get(), "custom_attributes", customAttributesData.get() );
```
Hey John,

can we continue the discussion about this? We currently use the `"custom_attributes"` attribute from Jabuka so set metadata and the way I implemented it here would stomp over that. That currently means we can't release a new Gaffer until we also fix Jabuka and release both at the same time.

Would you object if we check whether that `StringVectorData` actually contains any data and only set the attribute if it does? We might also want to print a warning if any stomping over actually happens?

Thanks,

Matti.

johnhaddon replied:
Can we not just support both? By initialising the array with any custom_attributes value we find in the parameters, and then appending the "header:" values?

mattigruener replied:
Yes, of course -.- See https://github.com/GafferHQ/gaffer/pull/1930
